### ISSUE TITLE
plugin:// search path url

### ISSUE THREAD

**Issue by jurialmunkey** [croneter/PlexKodiConnect#1103]
Hi @croneter 

Is there a plugin:// path url that a search query can be added to that will return a plugin container of matches? I have a user attempting to create a TMDbHelper player file for plexkodiconnect: https://github.com/jurialmunkey/plugin.video.themoviedb.helper/issues/50

The TMDbHelper player function essentially gets the results of a plugin path and then walks through the path to locate the correct item by matching ListItem.InfoLabels with the info it is looking for.

For instance, to play a movie with Netflix, I have two steps:
```
"plugin://plugin.video.netflix/directory/search/search/{title}/",
{"title": "{title}", "year": "{year}"}
```
That opens the plugin search path with a query for the {title} and then looks for an item with ListItem.Title and ListItem.Year that matches the {title} {year} being searched for. If it finds a match, it plays the item and if it fails it presents the user with other options.

Having a working plexkodiconnect player would be fantastic as it would allow users of my plugin to integrate their local library with plexkodiconnect and other plugins that they use (e.g. Netflix). OpenMeta has a player file but it uses a hack that inputs characters into the keyboard dialog, which is a method that won't be permitted in official Kodi repo and so is something that I won't add to TMDbHelper.

Since I'm not a plexkodiconnect user, any insight you can offer on usable search plugin paths would be greatly appreciated!

**Comment by croneter** [comment]
Hi @jurialmunkey !

 This sounds really promising, but PKC does not yet support what you're looking for. I think I could implement this rather quickly as the Plex api does have a search function. 

I need more info:
* How exactly would the search string look like?
* What attributes can one search for? (media type, title, genre, length of video,...?
* The actual searching would be done by PKC, correct, not your addon? Results amount to potentially thousands of items, which kills Kodi if we're talking listitems 
* What results can be returned? Movies, episodes, TV shows, seasons, even an individual actor? 
* How would PKC return 0, 1, n results? I hate listitems - could you point me to some working code?

PKC syncs content to the Kodi database. Is it not possible to hit Kodi directly with a query? 

The user can choose to NOT sync certain Plex libraries. Implementing the above would have the benefit that even for un-synced search results, playback via PKC would be possible

**Comment by jurialmunkey** [comment]
**A little bit of context/background**

TMDbHelper allows users to browse TMDb and Trakt via the APIs. If the user clicks a movie/episode, the player function is run. Skins can also call the player function onclick by passing a tmdb/tvdb/imdb ID or a title query string of the item they want to play. The player function does the following:

- Find the TMDb ID for the item and look-up it's details from the TMDb API
- Search the local library for a matching item and play if found
- If it can't find a local item, it attempts to play using an external player (either user's configured default or if that fails, then presents a list of options to choose from).
- It converts the TMDb data to a dictionary which is used to format the plugin's search path. Available keys are in the wiki: https://github.com/jurialmunkey/plugin.video.themoviedb.helper/wiki/Player-Function#available-keys
- The formatted plugin path is then opened using JSON-RPC getFiles method
- The player then looks for a listitem with infolabels that match the defined key values. Multiple steps can be defined to open a series of folders (e.g. plugin path to search for a tvshow -> open matching tvshow -> open matching season -> play matching episode)

**So in answer to your questions**
> How exactly would the search string look like?

Since I format the string using a dictionary, the syntax of the plugin path doesn't matter as long as it accepts a query string of some kind. For instance, I could work with either of the following styles because I can just replace "Mr Robot" with {showname} in the player file which the string formatter will then replace with the appropriate tvshow title.
```
plugin://plugin.video.plexkodiconnect/?info=search&type=tv&query=Mr Robot
plugin://plugin.video.plexkodiconnect/tv/search/Mr Robot
```

> What attributes can one search for? (media type, title, genre, length of video,...?

Anything really. It's ideal if the result listitems have a unique identifier such as imdb/tvdb/tmdb id, as that will allow for a perfect match but I can work with any info (or combo of info - e.g. title + year).

> The actual searching would be done by PKC, correct, not your addon? Results amount to potentially thousands of items, which kills Kodi if we're talking listitems

Yes that's correct. However, if a mediatype can be specified, then it shouldn't return more than a handful of results. For instance, if I'm searching PKC for the movie "Alien", I would also get results for "Aliens" "Alien 3" "Alien Resurrection" "Alien Covenant" etc. - but definitely not 1000s of results. Plus that's with a very generic title - most movie titles will only have one match.

> What results can be returned? Movies, episodes, TV shows, seasons, even an individual actor?

All I need is a search path for movies and a search path for tvshows. When playing an episode it is easier and more efficient for TMDbHelper to locate the tvshow and then walk the path looking for matching season/episode rather than getting a list of episodes and trying to find the correct one.

> How would PKC return 0, 1, n results? I hate listitems - could you point me to some working code?

The results just need to be in a normal plugin directory like you would make any other plugin directory. For 0 results you just end the directory without adding any listitems. 

> PKC syncs content to the Kodi database. Is it not possible to hit Kodi directly with a query?

TMDbHelper *always* attempts to play content in the local db first. I assume that the use case is for Plex libraries that the user doesn't sync.

The other advantage is that users can add TMDbHelper items to their kodi library and then use TMDbHelper to decide what source to play from

**Comment by jurialmunkey** [comment]
Also, as far as I understand there is also a search function built into PKC that can be accessed with:
```
plugin://plugin.video.plexkodiconnect/?mode=browseplex&key=%2fhubs%2fsearch%3ftype%3d2&prompt=Search+TVShow
```

The issue is that that will pop-up a keyboard dialog to enter the query. All I need is a way to pass the query directly to the plugin path and bypass the keyboard. e.g. Something like:
```
plugin://plugin.video.plexkodiconnect/?mode=browseplex&key=%2fhubs%2fsearch%3ftype%3d2&prompt=Search+TVShow&query=Mr Robot
```

**Comment by webmaher** [comment]
Hi @croneter
is any news about this:

> PKC does not yet support what you're looking for. I think I could implement this rather quickly as the Plex api does have a search function.

Thanks for your PKC which is work great :-)

**Comment by croneter** [comment]
@webmaher , @jurialmunkey 
Try PKC **BETA** version 2.10.9. Your search string would need to look like this:
```
plugin://plugin.video.plexkodiconnect?mode=search&query=YOUR SEARCH STRING HERE
```

**Comment by jurialmunkey** [comment]
@croneter - Excellent! Thanks 😁 

One question:
Is there a way to search by a specific mediatype, e.g. `movies` or `tvshows` or is it only possible to search with all mediatypes mixed together?

**Comment by croneter** [comment]
@jurialmunkey not as far as I know.

Background: the search query is simply forwarded to the PMS and then the result is translated to Kodi elements, be it episode, set, movie or actor

**Comment by jurialmunkey** [comment]
@croneter - Ah okay. Shouldn't be too much of a problem. 

Does it return the parent tvshow for a search? Or will it only return episodes?

**Comment by croneter** [comment]
Plex will also return shows if the query matches.

**Comment by jurialmunkey** [comment]
> 
> 
> Plex will also return shows if the query matches.

Okay awesome, thanks so much! 

I think my standard approach to the player file should work then. I'll let you know if I hit any roadblocks -- I'm pretty sure everything should work fine though. Still, I'm practically going in blind since I don't use Plex myself and need to rely on others to test!

**Comment by webmaher** [comment]
Big thanks @croneter ...just amazing work for you :-)

**Comment by croneter** [comment]
Available for everyone with version 2.10.12
